### PR TITLE
Reintroduce RematerializeParallelOpsPass as preprocessing

### PR DESCRIPTION
**Pull request by qedawkins**
This was removed in #14453. This reintroduces the pass as a preprocessing option to allow user application on pre-formed dispatches.

### CODE REVIEW

File: compiler/src/iree/compiler/Preprocessing/Common/RematerializeParallelOps.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+namespace {
+
+static bool isScalarOrTensorOfSizeOne(Type t) {
+  if (auto tensorType = dyn_cast<RankedTensorType>(t)) {
+    return tensorType.hasStaticShape() && tensorType.getNumElements() == 1;
+  }
+  return t.isIntOrIndexOrFloat();
+}
+
+/// Merge elementwise operations into their consumers.
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Preprocessing/Common/RematerializeParallelOps.cpp]: Maybe rename this to `RematerializeParallelOpsPattern` and add some comments to say this is rematerializing all parallel elementwise operations into its users within a `flow.dispatch.region`.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks! THis looks good!

**Comment by qedawkins** [comment]
hmmm the CI failure seems unrelated
```
FAILED: e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb 
cd /work/build-e2e-test-artifacts/tests/e2e/test_artifacts && /work/full-build-dir/install/bin/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda --iree-input-type=none --iree-hal-cuda-llvm-target-arch=sm_80 /work/build-e2e-test-artifacts/e2e_test_artifacts/model_Unet2dPT.mlir -o /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb --iree-hal-executable-object-search-path=\"/work/build-e2e-test-artifacts\"
iree-compile: /work/third_party/llvm-project/llvm/lib/Support/raw_ostream.cpp:114: void llvm::raw_ostream::SetBufferAndMode(char *, size_t, llvm::raw_ostream::BufferKind): Assertion `GetNumBytesInBuffer() == 0 && "Current buffer is non-empty!"' failed.
```

**Comment by pzread** [comment]
> hmmm the CI failure seems unrelated
> 
> ```
> FAILED: e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb 
> cd /work/build-e2e-test-artifacts/tests/e2e/test_artifacts && /work/full-build-dir/install/bin/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda --iree-input-type=none --iree-hal-cuda-llvm-target-arch=sm_80 /work/build-e2e-test-artifacts/e2e_test_artifacts/model_Unet2dPT.mlir -o /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb --iree-hal-executable-object-search-path=\"/work/build-e2e-test-artifacts\"
> iree-compile: /work/third_party/llvm-project/llvm/lib/Support/raw_ostream.cpp:114: void llvm::raw_ostream::SetBufferAndMode(char *, size_t, llvm::raw_ostream::BufferKind): Assertion `GetNumBytesInBuffer() == 0 && "Current buffer is non-empty!"' failed.
> ```

It's due to #14417, the fix is merged in the upstream but we need to wait for the integration. Re-run the failed job should work for now (it's flaky)

**Comment by qedawkins** [comment]
> > hmmm the CI failure seems unrelated
> > ```
> > FAILED: e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb 
> > cd /work/build-e2e-test-artifacts/tests/e2e/test_artifacts && /work/full-build-dir/install/bin/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda --iree-input-type=none --iree-hal-cuda-llvm-target-arch=sm_80 /work/build-e2e-test-artifacts/e2e_test_artifacts/model_Unet2dPT.mlir -o /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_Unet2dPT_linalg___cuda-sm_80-linux_gnu-cuda__default-flags_/module.vmfb --iree-hal-executable-object-search-path=\"/work/build-e2e-test-artifacts\"
> > iree-compile: /work/third_party/llvm-project/llvm/lib/Support/raw_ostream.cpp:114: void llvm::raw_ostream::SetBufferAndMode(char *, size_t, llvm::raw_ostream::BufferKind): Assertion `GetNumBytesInBuffer() == 0 && "Current buffer is non-empty!"' failed.
> > ```
> 
> It's due to #14417, the fix is merged in the upstream but we need to wait for the integration. Re-run the failed job should work for now (it's flaky)

I see, thanks!

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by qedawkins** [comment]
I would like to land this but I am still waiting for it to reserve an A100, although these changes are unrelated to anything checked in the A100 test. Is there a way around the CI in this case?

**Comment by ScottTodd** [comment]
> I would like to land this but I am still waiting for it to reserve an A100, although these changes are unrelated to anything checked in the A100 test. Is there a way around the CI in this case?

Probably simplest to just have an admin (like myself) merge without waiting for checks. You could also put `ci-skip: test_a100` in the PR description (https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/contributing.md#ci-behavior-manipulation), but I don't think it's worth re-triggering an entirely new CI run for just that.

**Comment by ScottTodd** [comment]
Oh, and we had been thinking about moving that a100 CI job to postsubmit due to the device shortages, but that would need a bit of CI refactoring.

**Comment by qedawkins** [comment]
> Oh, and we had been thinking about moving that a100 CI job to postsubmit due to the device shortages, but that would need a bit of CI refactoring.

Would this mean making it off by default, but still an opt-in for presubmit?

**Comment by ScottTodd** [comment]
Ideally, yes. `ci-extra: test_a100` would opt-in for presubmit.

**Comment by qedawkins** [comment]
+1 to that change then :D

**Pull request merged**